checks: code from fulcrum demo

asah opened the pull request.

dimaryaz commented on quilt/tools/build.py:
```diff
@@ -144,29 +172,49 @@ def _callback(count):
+        fd = open(checks_path)
+        docs = yaml.load_all(fd)
+        contents = next(docs, None)['contents'] # leave other dicts in the generator
+        if not isinstance(contents, dict):
```
There might be all kinds of other errors, too - e.g., `next` could return `None` or `contents` might be completely missing. (You can use `.get('contents')` to deal with that).

Unfortunately, there doesn't seem to be an easy way to validate yaml.

asah replied:
???  sorry, I'm unclear what you want?

dimaryaz commented on quilt/tools/build.py:
```diff
@@ -178,10 +226,28 @@ def build_package_from_contents(username, package, build_dir, data):
         raise BuildException("HDF5 format is no longer supported; please use PARQUET instead.")
 
+    # inline checks take precedence
+    if 'checks' in build_data:
```
`checks_contents.update(build_data.get('checks', {}))`

asah replied:
DONE

dimaryaz commented on quilt/tools/build.py:
```diff
@@ -178,10 +226,28 @@ def build_package_from_contents(username, package, build_dir, data):
+    if dryrun:
+        from .core import RootNode
+        from .package import Package
+        newpackage = Package(
```
Might be cleaner to add a dry-run option to `store.create_package`

asah replied:
DONE

dimaryaz commented on quilt/tools/check_functions.py:
```diff
@@ -0,0 +1,103 @@
+import numpy
+import pytest
+
+filename = None
```
Globals?

asah replied:
COMMENTED - needed for checks: syntax...

dimaryaz commented on quilt/tools/check_functions.py:
```diff
@@ -0,0 +1,103 @@
+                minval = col.min()
+            if maxval is None:
+                maxval = col.max()
+            if lambda_or_name in ["mean", "avg"]:
```
A little messy... Maybe define a global `dict` of names to lambdas?

asah replied:
DONE

dimaryaz commented on quilt/tools/check_functions.py:
```diff
@@ -0,0 +1,103 @@
+
+filename = None
+data = None
+env = 'default'
```
What are envs? I couldn't quite make sense of them.

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -648,6 +652,18 @@ def install(package, hash=None, version=None, tag=None, force=False):
         )
         pkghash = response.json()['hash']
+    elif len(hash) < 64:
+        print('matching short hash against logs...')
```
Messages should be capitalized

asah replied:
DONE: removed

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -648,6 +652,18 @@ def install(package, hash=None, version=None, tag=None, force=False):
+            )
+        )
+        for entry in reversed(response.json()['logs']):
+            if entry['hash'][:len(hash)] == hash:
```
You can also just use `.endswith`

asah replied:
DONE: nice

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -648,6 +652,18 @@ def install(package, hash=None, version=None, tag=None, force=False):
+        )
+        for entry in reversed(response.json()['logs']):
+            if entry['hash'][:len(hash)] == hash:
+                pkghash = entry['hash']
```
`break` after you found it?

Also, `else` after the `for` loop to handle hash not found

asah replied:
DONE - good catch

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -648,6 +652,18 @@ def install(package, hash=None, version=None, tag=None, force=False):
             )
         )
         pkghash = response.json()['hash']
+    elif len(hash) < 64:
```
Probably want to check it's not empty

asah replied:
DONE

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -721,11 +737,94 @@ def install(package, hash=None, version=None, tag=None, force=False):
+    environments = None
+    with open('build.yml') as fd:
+        buildfile = next(yaml.load_all(fd), None)
+        environments = buildfile.get('environments')
```
Use `buildfile.get('environments', {})` - then you don't have to check for `None`. Actually, may want to check that it's a `dict`

asah replied:
DONE

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -721,11 +737,94 @@ def install(package, hash=None, version=None, tag=None, force=False):
+def _setup_env(env, files):
+    """ process data distribution. """
+    # TODO: build.yml is not saved in the package system, so re-load it here
+    environments = None
```
Not needed

asah replied:
DONE

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -721,11 +737,94 @@ def install(package, hash=None, version=None, tag=None, force=False):
+    dataset = environment.get('dataset')
+    for key, val in files.items():
+        # TODO: debug mode, where we can see which files were skipped
+        if type(val) == pd.core.frame.DataFrame:
```
Use `isinstance`, to allow subclasses. Also, `pd.DataFrame` is the same thing.

asah replied:
DONE

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -721,11 +737,94 @@ def install(package, hash=None, version=None, tag=None, force=False):
+        if type(val) == pd.core.frame.DataFrame:
+            before_len = len(val)
+            res = exec_yaml_python(dataset, val, key, '('+key+')')
+            if res == False:
```
`if not res:`

asah replied:
DONE

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -721,11 +737,94 @@ def install(package, hash=None, version=None, tag=None, force=False):
+    import importlib
+    install(pkg, force=True)
+    user, module = pkg.split('/')
+    module = importlib.import_module("quilt.data.{}.{}".format(user, module))
```
Take a look at the code in `data.py` and use store directly, instead of importing the module.

asah replied:
DONE.  nice!

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -721,11 +737,94 @@ def install(package, hash=None, version=None, tag=None, force=False):
+    resdict = load_to_dict(pkg, force, env, **kwargs)
+    for key, val in resdict.items():
+        # HACK: use the object store?  how to garbage collect?
+        newfn = '/tmp/'+getpass.getuser()+"-"+key
```
Use `tempfile` at the very least.

asah replied:
DONE

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -721,11 +737,94 @@ def install(package, hash=None, version=None, tag=None, force=False):
+    files = _setup_env(env, files)
+    return files
+
+def load_to_files(pkg, force=False, format='tsv', env='default', **kwargs):
```
What exactly does this do?

asah replied:
REMOVED - see below

dimaryaz commented on quilt/tools/core.py:
```diff
@@ -2,8 +2,25 @@
 import struct
 
+# for check functions
+import pandas
```
`core.py` is actually shared with the backend (by copying the file... yeah, we need to fix that). So don't add client-specific code here.

asah replied:
COMMENTED - added a comment about this.  But I also left the code because the checks will eventually need to run server-side (in a container...)

asah commented:
Killer code review!  I'm on it, will address each issue.

(via cellphone, random spelling ahead)

On Sep 6, 2017 12:20 PM, "Dima Ryazanov" <notifications@github.com> wrote:

> *@dimaryaz* requested changes on this pull request.
>
> Looks like you forgot to add checks_simple.yml.
>
> Also, run pylint. Don't worry about fixing everything (missing doc
> strings, short variable names, whatever) - but it often points out actual
> bugs.
> ------------------------------
>
> In quilt/test/test_checks.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137340027>:
>
> > +Test the data checking / linting system
> +"""
> +import os
> +import re
> +
> +from six import assertRaisesRegex
> +import yaml
> +import pytest
> +
> +from ..tools.package import Package
> +from ..tools import build, command
> +from .utils import QuiltTestCase
> +
> +def read_yml_file(fn):
> +    mydir = os.path.dirname(__file__)
> +    path = os.path.join(mydir, '.')
>
> Not needed - it's just '.'
> ------------------------------
>
> In quilt/test/test_checks.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137341214>:
>
> > +
> +def read_yml_file(fn):
> +    mydir = os.path.dirname(__file__)
> +    path = os.path.join(mydir, '.')
> +    filepath = os.path.join(path, fn)
> +    with open(filepath) as fd:
> +        return next(yaml.load_all(fd), None)
> +
> +class ChecksTest(QuiltTestCase):
> +
> +    def setUp(self):
> +        self.checks_data = read_yml_file('checks_simple.yml')
> +        self.checks_contents = self.checks_data['contents']
> +        self.build_data = read_yml_file('build_simple_checks.yml')
> +        self.build_contents = self.build_data['contents']
> +        return super(QuiltTestCase, self).setUp()
>
> Don't need return: setUp doesn't return anything.
>
> Also, call super as the first line, just in case. (QuiltTestCase.setUp
> does a bunch of important things like mocking network requests - so that
> needs to happen before any test code runs.)
> ------------------------------
>
> In quilt/test/test_checks.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137341260>:
>
> > +    mydir = os.path.dirname(__file__)
> +    path = os.path.join(mydir, '.')
> +    filepath = os.path.join(path, fn)
> +    with open(filepath) as fd:
> +        return next(yaml.load_all(fd), None)
> +
> +class ChecksTest(QuiltTestCase):
> +
> +    def setUp(self):
> +        self.checks_data = read_yml_file('checks_simple.yml')
> +        self.checks_contents = self.checks_data['contents']
> +        self.build_data = read_yml_file('build_simple_checks.yml')
> +        self.build_contents = self.build_data['contents']
> +        return super(QuiltTestCase, self).setUp()
> +
> +    def tearDown(self):
>
> Not needed
> ------------------------------
>
> In quilt/test/test_checks.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137341948>:
>
> > +        self.build_contents = self.build_data['contents']
> +        return super(QuiltTestCase, self).setUp()
> +
> +    def tearDown(self):
> +        return super(QuiltTestCase, self).tearDown()
> +
> +    def build_success(self, check):
> +        self.build_contents['foo']['checks'] = check
> +        build.build_package_from_contents(
> +            'foo', 'bar', self._old_dir, self.build_data, self.checks_contents,
> +            dryrun=True)
> +
> +    def build_fail(self, check, regexp=None):
> +        if regexp is None:
> +            regexp = "Data check failed: %s" % (check)
> +        try:
>
> I think with assertRaisesRegex below does exactly this? You shouldn't
> need the try block anymore.
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137342436>:
>
> > @@ -37,12 +35,27 @@ def _pythonize_name(name):
>          raise BuildException("Unable to determine a Python-legal name for %r" % name)
>      return safename
>
> -def _build_node(build_dir, package, name, node, fmt, target='pandas'):
> +def _run_checks(dataframe, checks, checks_contents, nodename, rel_path, target, env='default'):
> +    print("Running data integrity checks...")
> +    checks_list = re.split(r'[,\s]+', checks.strip())
> +    unknown_checks = set(checks_list) - set(checks_contents.keys())
> +    if len(unknown_checks) > 0:
> +        raise BuildException("Unknown check(s) '%s' for %s @ %s" %
> +                             (", ".join(list(unknown_checks)), rel_path, target))
> +    for check in checks_list:
> +        res = exec_yaml_python(checks_contents[check], dataframe, nodename, rel_path, target)
> +        if res == False:
>
> if not res:
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137342672>:
>
> > -def _build_node(build_dir, package, name, node, fmt, target='pandas'):
> +def _run_checks(dataframe, checks, checks_contents, nodename, rel_path, target, env='default'):
> +    print("Running data integrity checks...")
> +    checks_list = re.split(r'[,\s]+', checks.strip())
> +    unknown_checks = set(checks_list) - set(checks_contents.keys())
> +    if len(unknown_checks) > 0:
> +        raise BuildException("Unknown check(s) '%s' for %s @ %s" %
> +                             (", ".join(list(unknown_checks)), rel_path, target))
> +    for check in checks_list:
> +        res = exec_yaml_python(checks_contents[check], dataframe, nodename, rel_path, target)
> +        if res == False:
> +            raise BuildException("Data check failed: %s on %s @ %s" % (
> +                check, rel_path, target))
> +
> +def _build_node(build_dir, package, name, node, format, target='pandas', checks_contents=None,
> +                dryrun=False, env='default'):
>
> Nit: I'd prefer dry_run
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137342877>:
>
> >          if transform == ID:
> -            print("Copying %s..." % path)
> -            package.save_file(path, name, rel_path)
> +            if checks:
> +                data = open(path, 'r').read()
>
> with open(...) - otherwise, the file stays open until it gets eventually
> garbage collected
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137343312>:
>
> >      """
>      Builds a package from a given Yaml file and installs it locally.
>
>      Returns the name of the package.
>      """
> +    def find(key, value):
> +        for k, v in value.items():
>
> Just use value.get(key)?
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137343800>:
>
> >      build_dir = os.path.dirname(yaml_path)
>      fd = open(yaml_path)
>      docs = yaml.load_all(fd)
> -    data = next(docs, None) # leave other dicts in the generator
> -    if not isinstance(data, dict):
> -        raise BuildException("Unable to parse YAML: %s" % yaml_path)
> -
> -    build_package_from_contents(username, package, build_dir, data)
> -
> -def build_package_from_contents(username, package, build_dir, data):
> -    """
> -    Builds package according to YAML node
> -    """
> -    contents = data.get('contents', {})
> +    build_data = next(docs, None) # leave other dicts in the generator
> +    if not isinstance(build_data, dict):
> +        raise BuildException("Unable to parse build YAML: %s" % yaml_path)
> +    if len(list(find('checks', build_data['contents']))) > 0 and 'checks' not in build_data:
>
> Something like if next(find(...)) and ... should work.
>
> But in general, if you want to check if a list lst is non-empty, use if
> lst rather than if len(lst) > 0
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137343926>:
>
> > -    build_package_from_contents(username, package, build_dir, data)
> -
> -def build_package_from_contents(username, package, build_dir, data):
> -    """
> -    Builds package according to YAML node
> -    """
> -    contents = data.get('contents', {})
> +    build_data = next(docs, None) # leave other dicts in the generator
> +    if not isinstance(build_data, dict):
> +        raise BuildException("Unable to parse build YAML: %s" % yaml_path)
> +    if len(list(find('checks', build_data['contents']))) > 0 and 'checks' not in build_data:
> +        checks_path = 'checks.yml'
> +
> +    checks_contents = None
> +    if checks_path is not None:
> +        fd = open(checks_path)
>
> Use with
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137344197>:
>
> > -    """
> -    Builds package according to YAML node
> -    """
> -    contents = data.get('contents', {})
> +    build_data = next(docs, None) # leave other dicts in the generator
> +    if not isinstance(build_data, dict):
> +        raise BuildException("Unable to parse build YAML: %s" % yaml_path)
> +    if len(list(find('checks', build_data['contents']))) > 0 and 'checks' not in build_data:
> +        checks_path = 'checks.yml'
> +
> +    checks_contents = None
> +    if checks_path is not None:
> +        fd = open(checks_path)
> +        docs = yaml.load_all(fd)
> +        contents = next(docs, None)['contents'] # leave other dicts in the generator
> +        if not isinstance(contents, dict):
>
> There might be all kinds of other errors, too - e.g., next could return
> None or contents might be completely missing. (You can use
> .get('contents') to deal with that).
>
> Unfortunately, there doesn't seem to be an easy way to validate yaml.
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137344746>:
>
> > @@ -178,10 +226,28 @@ def build_package_from_contents(username, package, build_dir, data):
>      if pkgformat is PackageFormat.HDF5:
>          raise BuildException("HDF5 format is no longer supported; please use PARQUET instead.")
>
> +    # inline checks take precedence
> +    if 'checks' in build_data:
>
> checks_contents.update(build_data.get('checks', {}))
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137345242>:
>
> >      store = PackageStore()
> -    newpackage = store.create_package(username, package)
> -    _build_node(build_dir, newpackage, '', contents, pkgformat)
> -    newpackage.save_contents()
> +    if dryrun:
> +        from .core import RootNode
> +        from .package import Package
> +        newpackage = Package(
>
> Might be cleaner to add a dry-run option to store.create_package
> ------------------------------
>
> In quilt/tools/check_functions.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137345660>:
>
> > @@ -0,0 +1,103 @@
> +import re
> +from pandas import DataFrame as df
> +import numpy
> +import pytest
> +
> +filename = None
>
> Globals?
> ------------------------------
>
> In quilt/tools/check_functions.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137346508>:
>
> > +
> +def check_column_valrange(colrx, minval=None, maxval=None, lambda_or_name=None, envs=None):
> +    if envs not in [None, 'default']: check_column_regexp(colrx, envs[env])
> +    for colname in list(data):
> +        if re.search(colrx, colname):
> +            if minval is None and maxval is None:
> +                raise CheckFunctionsException(
> +                    'check_column_valrange() requires minval or maxval')
> +            col = data[colname]
> +            if callable(lambda_or_name):
> +                col = lambda_or_name(col)
> +            if minval is None:
> +                minval = col.min()
> +            if maxval is None:
> +                maxval = col.max()
> +            if lambda_or_name in ["mean", "avg"]:
>
> A little messy... Maybe define a global dict of names to lambdas?
> ------------------------------
>
> In quilt/tools/check_functions.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137346754>:
>
> > @@ -0,0 +1,103 @@
> +import re
> +from pandas import DataFrame as df
> +import numpy
> +import pytest
> +
> +filename = None
> +data = None
> +env = 'default'
>
> What are envs? I couldn't quite make sense of them.
> ------------------------------
>
> In quilt/tools/command.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137347383>:
>
> > @@ -648,6 +652,18 @@ def install(package, hash=None, version=None, tag=None, force=False):
>              )
>          )
>          pkghash = response.json()['hash']
> +    elif len(hash) < 64:
> +        print('matching short hash against logs...')
>
> Messages should be capitalized
> ------------------------------
>
> In quilt/tools/command.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137347568>:
>
> > @@ -648,6 +652,18 @@ def install(package, hash=None, version=None, tag=None, force=False):
>              )
>          )
>          pkghash = response.json()['hash']
> +    elif len(hash) < 64:
> +        print('matching short hash against logs...')
> +        response = session.get(
> +            "{url}/api/log/{owner}/{pkg}/".format(
> +                url=QUILT_PKG_URL,
> +                owner=owner,
> +                pkg=pkg
> +            )
> +        )
> +        for entry in reversed(response.json()['logs']):
> +            if entry['hash'][:len(hash)] == hash:
>
> You can also just use .endswith
> ------------------------------
>
> In quilt/tools/command.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137347695>:
>
> > @@ -648,6 +652,18 @@ def install(package, hash=None, version=None, tag=None, force=False):
>              )
>          )
>          pkghash = response.json()['hash']
> +    elif len(hash) < 64:
> +        print('matching short hash against logs...')
> +        response = session.get(
> +            "{url}/api/log/{owner}/{pkg}/".format(
> +                url=QUILT_PKG_URL,
> +                owner=owner,
> +                pkg=pkg
> +            )
> +        )
> +        for entry in reversed(response.json()['logs']):
> +            if entry['hash'][:len(hash)] == hash:
> +                pkghash = entry['hash']
>
> break after you found it?
>
> Also, else after the for loop to handle hash not found
> ------------------------------
>
> In quilt/tools/command.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137348261>:
>
> > @@ -648,6 +652,18 @@ def install(package, hash=None, version=None, tag=None, force=False):
>              )
>          )
>          pkghash = response.json()['hash']
> +    elif len(hash) < 64:
>
> Probably want to check it's not empty
> ------------------------------
>
> In quilt/tools/command.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137348726>:
>
> > @@ -721,11 +737,94 @@ def install(package, hash=None, version=None, tag=None, force=False):
>
>      pkgobj.save_contents()
>
> +def _setup_env(env, files):
> +    """ process data distribution. """
> +    # TODO: build.yml is not saved in the package system, so re-load it here
> +    environments = None
> +    with open('build.yml') as fd:
> +        buildfile = next(yaml.load_all(fd), None)
> +        environments = buildfile.get('environments')
>
> Use buildfile.get('environments', {}) - then you don't have to check for
> None. Actually, may want to check that it's a dict
> ------------------------------
>
> In quilt/tools/command.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137348754>:
>
> > @@ -721,11 +737,94 @@ def install(package, hash=None, version=None, tag=None, force=False):
>
>      pkgobj.save_contents()
>
> +def _setup_env(env, files):
> +    """ process data distribution. """
> +    # TODO: build.yml is not saved in the package system, so re-load it here
> +    environments = None
>
> Not needed
> ------------------------------
>
> In quilt/tools/command.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137349198>:
>
> > +    with open('build.yml') as fd:
> +        buildfile = next(yaml.load_all(fd), None)
> +        environments = buildfile.get('environments')
> +    if env != 'default' and (environments is None or env not in environments):
> +        raise CommandException(
> +            "environment %s not found in environments: section of build.yml" % env)
> +    if environments is None:
> +        return files
> +    if env == 'default' and 'default' not in environments:
> +        return files
> +    print('processing environment %s: checking data...' % (env))
> +    environment = environments[env]
> +    dataset = environment.get('dataset')
> +    for key, val in files.items():
> +        # TODO: debug mode, where we can see which files were skipped
> +        if type(val) == pd.core.frame.DataFrame:
>
> Use isinstance, to allow subclasses. Also, pd.DataFrame is the same thing.
> ------------------------------
>
> In quilt/tools/command.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137349262>:
>
> > +    if env != 'default' and (environments is None or env not in environments):
> +        raise CommandException(
> +            "environment %s not found in environments: section of build.yml" % env)
> +    if environments is None:
> +        return files
> +    if env == 'default' and 'default' not in environments:
> +        return files
> +    print('processing environment %s: checking data...' % (env))
> +    environment = environments[env]
> +    dataset = environment.get('dataset')
> +    for key, val in files.items():
> +        # TODO: debug mode, where we can see which files were skipped
> +        if type(val) == pd.core.frame.DataFrame:
> +            before_len = len(val)
> +            res = exec_yaml_python(dataset, val, key, '('+key+')')
> +            if res == False:
>
> if not res:
> ------------------------------
>
> In quilt/tools/command.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137349585>:
>
> > +        if type(val) == pd.core.frame.DataFrame:
> +            before_len = len(val)
> +            val['.qchash'] = val.apply(lambda x: abs(hash(tuple(x))), axis = 1)
> +            res = exec_yaml_python(dataset, val, key, '('+key+')')
> +            if res == False:
> +                raise BuildException("Data check failed: %s on %s @ %s" % (
> +                    check, rel_path, target))
> +            print('%s: %s=>%s recs' % (key, before_len, len(qc.data)))
> +            files[key] = qc.data
> +    return files
> +
> +def load_to_dict(pkg, force=False, env='default', **kwargs):
> +    import importlib
> +    install(pkg, force=True)
> +    user, module = pkg.split('/')
> +    module = importlib.import_module("quilt.data.{}.{}".format(user, module))
>
> Take a look at the code in data.py and use store directly, instead of
> importing the module.
> ------------------------------
>
> In quilt/tools/command.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137350231>:
>
> > +    return files
> +
> +def load_to_dict(pkg, force=False, env='default', **kwargs):
> +    import importlib
> +    install(pkg, force=True)
> +    user, module = pkg.split('/')
> +    module = importlib.import_module("quilt.data.{}.{}".format(user, module))
> +    files = dict([ (key, getattr(module, val)()) for key, val in kwargs.items() ])
> +    files = _setup_env(env, files)
> +    return files
> +
> +def load_to_files(pkg, force=False, format='tsv', env='default', **kwargs):
> +    resdict = load_to_dict(pkg, force, env, **kwargs)
> +    for key, val in resdict.items():
> +        # HACK: use the object store?  how to garbage collect?
> +        newfn = '/tmp/'+getpass.getuser()+"-"+key
>
> Use tempfile at the very least.
> ------------------------------
>
> In quilt/tools/command.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137350267>:
>
> > +                raise BuildException("Data check failed: %s on %s @ %s" % (
> +                    check, rel_path, target))
> +            print('%s: %s=>%s recs' % (key, before_len, len(qc.data)))
> +            files[key] = qc.data
> +    return files
> +
> +def load_to_dict(pkg, force=False, env='default', **kwargs):
> +    import importlib
> +    install(pkg, force=True)
> +    user, module = pkg.split('/')
> +    module = importlib.import_module("quilt.data.{}.{}".format(user, module))
> +    files = dict([ (key, getattr(module, val)()) for key, val in kwargs.items() ])
> +    files = _setup_env(env, files)
> +    return files
> +
> +def load_to_files(pkg, force=False, format='tsv', env='default', **kwargs):
>
> What exactly does this do?
> ------------------------------
>
> In quilt/tools/core.py
> <https://github.com/quiltdata/quilt/pull/162#discussion_r137350613>:
>
> > @@ -2,8 +2,25 @@
>  import hashlib
>  import struct
>
> +# for check functions
> +import pandas
>
> core.py is actually shared with the backend (by copying the file... yeah,
> we need to fix that). So don't add client-specific code here.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/quiltdata/quilt/pull/162#pullrequestreview-60994039>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIi4u5WktIp_WHwVz0_W8y6-Pep7WMCks5sfuNigaJpZM4PN8cU>
> .
>

asah commented on quilt/test/test_checks.py:
```diff
@@ -0,0 +1,108 @@
+
+def read_yml_file(fn):
+    mydir = os.path.dirname(__file__)
+    path = os.path.join(mydir, '.')
```
DONE: good catch, weird that I missed this.

asah commented on quilt/test/test_checks.py:
```diff
@@ -0,0 +1,108 @@
+        self.checks_contents = self.checks_data['contents']
+        self.build_data = read_yml_file('build_simple_checks.yml')
+        self.build_contents = self.build_data['contents']
+        return super(QuiltTestCase, self).setUp()
```
DONE - thx

asah commented on quilt/test/test_checks.py:
```diff
@@ -0,0 +1,108 @@
+    def build_fail(self, check, regexp=None):
+        if regexp is None:
+            regexp = "Data check failed: %s" % (check)
+        try:
```
COMMENTED: actually, this is a weird case where build_success() is supposed to fail in a particular way, but I also want to catch the case where build_success() accidentally succeeds and produce an easy-to-understand error.

asah commented on quilt/tools/build.py:
```diff
@@ -37,12 +35,27 @@ def _pythonize_name(name):
+                             (", ".join(list(unknown_checks)), rel_path, target))
+    for check in checks_list:
+        res = exec_yaml_python(checks_contents[check], dataframe, nodename, rel_path, target)
+        if res == False:
```
DONE: yeah, I suppose this is better... I was thinking of the case where it returns None... I added an explicit check for None

asah commented on quilt/tools/build.py:
```diff
@@ -37,12 +35,27 @@ def _pythonize_name(name):
+                check, rel_path, target))
+
+def _build_node(build_dir, package, name, node, format, target='pandas', checks_contents=None,
+                dryrun=False, env='default'):
```
DONE: good catch

asah commented on quilt/tools/build.py:
```diff
@@ -63,9 +76,17 @@ def _build_node(build_dir, package, name, node, fmt, target='pandas'):
-            print("Copying %s..." % path)
-            package.save_file(path, name, rel_path)
+            if checks:
+                data = open(path, 'r').read()
```
DONE: sorry good catch

asah commented on quilt/tools/build.py:
```diff
@@ -144,29 +172,49 @@ def _callback(count):
     Returns the name of the package.
     """
+    def find(key, value):
+        for k, v in value.items():
```
COMMENTED: finds all nodes transitively via yield

asah commented on quilt/tools/build.py:
```diff
@@ -144,29 +172,49 @@ def _callback(count):
+    build_data = next(docs, None) # leave other dicts in the generator
+    if not isinstance(build_data, dict):
+        raise BuildException("Unable to parse build YAML: %s" % yaml_path)
+    if len(list(find('checks', build_data['contents']))) > 0 and 'checks' not in build_data:
```
NOCHG: though I know it's more pythonic, it's harder for multi-language programmers to read so I like to be explicit.  ok?

dimaryaz replied:
I think we should actually use the Python conventions:
- Doesn't make sense to optimize for people who are not familiar with the language
- `list(...)` is actually inefficient here: it requires finding all results, even though the first one would be sufficient.

dimaryaz replied:
That said... I'm not sure I like the logic. It uses `checks.yml` - except when there's at least one `checks` field in the build file?

asah replied:
DONE re list.
COMMENT ADDED re logic -- inlined checks (in build.yml) should override external checks.yml -- otherwise, you can't safely create libraries of checks.

asah commented on quilt/tools/build.py:
```diff
@@ -144,29 +172,49 @@ def _callback(count):
+
+    checks_contents = None
+    if checks_path is not None:
+        fd = open(checks_path)
```
DONE

asah commented:
checks_simple.yml added
pylint run & fixed, also extended pylintrc

dimaryaz commented on quilt/tools/check_functions.py:
```diff
@@ -0,0 +1,112 @@
+from pandas import DataFrame as df
+import numpy
+
+# these are defined as globals so importing the library results in clean syntax
```
You should instead pass dicts of locals and globals to `eval` and `exec`

asah replied:
DONE

dimaryaz commented on quilt/tools/build.py:
```diff
@@ -135,38 +167,59 @@ def _callback(count):
     """
+    def find(key, value):
+        """find all nodes transitively"""
+        for k, v in value.items():
```
Use `iteritems` from `six`

asah replied:
DONE re list()...
NOCHG re iteritems - I need transitive iteration, iteritems() is shallow.

dimaryaz replied:
What do you mean by transitive iteration? `six.iteritems` should be the same as `iteritems` in Python2 and `items` in Python3.

asah replied:
DONE: haha I thought you meant all of find() not just this one line.

dimaryaz commented on quilt/tools/build.py:
```diff
@@ -37,89 +35,123 @@ def _pythonize_name(name):
+    print("Running data integrity checks...")
+    checks_list = re.split(r'[,\s]+', checks.strip())
+    unknown_checks = set(checks_list) - set(checks_contents.keys())
+    if len(unknown_checks) > 0:
```
`if unknown_checks`

asah replied:
DONE

dimaryaz commented on quilt/tools/build.py:
```diff
@@ -37,89 +35,123 @@ def _pythonize_name(name):
+    _ = env  # TODO: env support for checks
+    print("Running data integrity checks...")
+    checks_list = re.split(r'[,\s]+', checks.strip())
+    unknown_checks = set(checks_list) - set(checks_contents.keys())
```
I _think_ `.keys()` is not needed

asah replied:
NOCHG:

>>> set([3,4,5]) - dict({})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for -: 'set' and 'dict'

dimaryaz replied:
I meant, still call `set` - just without the `.keys()`:

```
>>> set([1, 2, 3]) - set({1: 'a', 2: 'b'})
{3}
```

asah replied:
DONE.  ah.

dimaryaz commented on quilt/tools/build.py:
```diff
@@ -9,19 +9,17 @@
-    Build-time exception class
-    """
-    pass
+# for use by check functions (eval)
```
Import them at the place you call `eval`, and pass them in as a locals or globals dict

asah replied:
DONE

asah commented:
Ok I'm on it, may need a few days since I'm traveling now.  Thx!!! For the
review.

(via cellphone, random spelling ahead)

On Sep 13, 2017 5:23 AM, "Dima Ryazanov" <notifications@github.com> wrote:

> *@dimaryaz* requested changes on this pull request.
> ------------------------------
>
> In quilt/tools/check_functions.py
> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r138413840>
> :
>
> > @@ -0,0 +1,112 @@
> +import re
> +import pandas as pd
> +from pandas import DataFrame as df
> +import numpy
> +
> +# these are defined as globals so importing the library results in clean syntax
>
> You should instead pass dicts of locals and globals to eval and exec
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r138414936>
> :
>
> >      build_dir = os.path.dirname(yaml_path)
>      fd = open(yaml_path)
>      docs = yaml.load_all(fd)
> -    data = next(docs, None) # leave other dicts in the generator
> -    if not isinstance(data, dict):
> -        raise BuildException("Unable to parse YAML: %s" % yaml_path)
> -
> -    build_package_from_contents(username, package, build_dir, data)
> -
> -def build_package_from_contents(username, package, build_dir, data):
> -    """
> -    Builds package according to YAML node
> -    """
> -    contents = data.get('contents', {})
> +    build_data = next(docs, None) # leave other dicts in the generator
> +    if not isinstance(build_data, dict):
> +        raise BuildException("Unable to parse build YAML: %s" % yaml_path)
> +    if len(list(find('checks', build_data['contents']))) > 0 and 'checks' not in build_data:
>
> I think we should actually use the Python conventions:
>
>    - Doesn't make sense to optimize for people who are not familiar with
>    the language
>    - list(...) is actually inefficient here: it requires finding all
>    results, even though the first one would be sufficient.
>
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r138416167>
> :
>
> >      build_dir = os.path.dirname(yaml_path)
>      fd = open(yaml_path)
>      docs = yaml.load_all(fd)
> -    data = next(docs, None) # leave other dicts in the generator
> -    if not isinstance(data, dict):
> -        raise BuildException("Unable to parse YAML: %s" % yaml_path)
> -
> -    build_package_from_contents(username, package, build_dir, data)
> -
> -def build_package_from_contents(username, package, build_dir, data):
> -    """
> -    Builds package according to YAML node
> -    """
> -    contents = data.get('contents', {})
> +    build_data = next(docs, None) # leave other dicts in the generator
> +    if not isinstance(build_data, dict):
> +        raise BuildException("Unable to parse build YAML: %s" % yaml_path)
> +    if len(list(find('checks', build_data['contents']))) > 0 and 'checks' not in build_data:
>
> That said... I'm not sure I like the logic. It uses checks.yml - except
> when there's at least one checks field in the build file?
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r138416245>
> :
>
> >      """
>      Builds a package from a given Yaml file and installs it locally.
>
>      Returns the name of the package.
>      """
> +    def find(key, value):
> +        """find all nodes transitively"""
> +        for k, v in value.items():
>
> Use iteritems from six
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r138416385>
> :
>
> > @@ -37,89 +35,123 @@ def _pythonize_name(name):
>          raise BuildException("Unable to determine a Python-legal name for %r" % name)
>      return safename
>
> -def _build_node(build_dir, package, name, node, fmt, target='pandas'):
> +def _run_checks(dataframe, checks, checks_contents, nodename, rel_path, target, env='default'):
> +    _ = env  # TODO: env support for checks
> +    print("Running data integrity checks...")
> +    checks_list = re.split(r'[,\s]+', checks.strip())
> +    unknown_checks = set(checks_list) - set(checks_contents.keys())
> +    if len(unknown_checks) > 0:
>
> if unknown_checks
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r138416442>
> :
>
> > @@ -37,89 +35,123 @@ def _pythonize_name(name):
>          raise BuildException("Unable to determine a Python-legal name for %r" % name)
>      return safename
>
> -def _build_node(build_dir, package, name, node, fmt, target='pandas'):
> +def _run_checks(dataframe, checks, checks_contents, nodename, rel_path, target, env='default'):
> +    _ = env  # TODO: env support for checks
> +    print("Running data integrity checks...")
> +    checks_list = re.split(r'[,\s]+', checks.strip())
> +    unknown_checks = set(checks_list) - set(checks_contents.keys())
>
> I *think* .keys() is not needed
> ------------------------------
>
> In quilt/tools/build.py
> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r138416552>
> :
>
> >  from .util import FileWithReadProgress
>
> -class BuildException(Exception):
> -    """
> -    Build-time exception class
> -    """
> -    pass
> +# for use by check functions (eval)
>
> Import them at the place you call eval, and pass them in as a locals or
> globals dict
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/quiltdata/quilt-compiler/pull/162#pullrequestreview-62205578>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIi4ma_2ujnF9i0fj3c0iBrREncCmHeks5shsB3gaJpZM4PN8cU>
> .
>

asah commented:
@dimaryaz changes made... thx!! again for the thorough review.

dimaryaz commented on quilt/tools/build.py:
```diff
@@ -135,38 +164,60 @@ def _callback(count):
+    if not isinstance(build_data, dict):
+        raise BuildException("Unable to parse build YAML: %s" % yaml_path)
+    # in-lined checks (in build.yml) take precedence over checks.yml
+    if find('checks', build_data['contents']) and 'checks' not in build_data:
```
Also, I think you do need to call `next(find(...))`. Even an empty generator evaluates to `True`.

asah replied:
DONE (as part of rewrite)

asah commented:
I rewrote/simplified build.build_package() and added some unit tests for it.

dimaryaz commented on quilt/tools/check_functions.py:
```diff
@@ -0,0 +1,112 @@
+                check(data[colname].isin(lambda_or_listexpr).all())
+
+VALRANGE_FUNCS = {
+    'mean':     lambda col, minval, maxval: check(minval <= col.mean() <= maxval),
```
I'd move `check` out of the function, and leave just the `lambda ...: minval <= ... <= maxval`.

(You could be really fancy make it a map of names to methods, e.g. `'mean': ColumnType.mean` - but I guess that doesn't work for `abs`? Anyways, just a thought - not sure that makes it better.)

asah replied:
DONE - also removed the minval/maxval

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -14,6 +14,9 @@
 import time
 from threading import Thread, Lock
+import yaml
+import getpass
```
Make it alphabetical? I think pylint complains about that (unless you disabled that?)

asah replied:
DONE

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -437,16 +443,16 @@ def _worker_thread():
-                                    ctx.compressed_read += count
-                                    original_read = ctx.compressed_read * original_size // compressed_size
+                                    Context.compressed_read += count
+                                    original_read = Content.compressed_read * original_size // compressed_size
```
Uh... `Context`, not `Content`?

This code is kinda impossible to unit test, unfortunately... Though I'd expect the linter to catch this.

asah replied:
DONE: good catch.  sigh, pylint FTL

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -648,6 +654,19 @@ def install(package, hash=None, version=None, tag=None, force=False):
         )
         pkghash = response.json()['hash']
+    elif len(hash) < 64:
+        assert len(hash) >= 6   # short is ok, but only to a point...
```
This can come from the command-line args. An error is ok - but make sure it doesn't assert on user input.

asah replied:
DONE

asah commented:
No worries from me!  I'm on it.

(via cellphone, random spelling ahead)

On Sep 20, 2017 4:37 PM, "Dima Ryazanov" <notifications@github.com> wrote:

> *@dimaryaz* requested changes on this pull request.
>
> Sorry, still found a couple problems.
>
> (This is just a really huge change - which is kinda scary...)
> ------------------------------
>
> In quilt/tools/check_functions.py
> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r140113756>
> :
>
> > +        if re.search(colrx, colname):
> +            if callable(lambda_or_listexpr):
> +                check(lambda_or_listexpr(data[colname]))
> +            else:
> +                check(data[colname].isin(lambda_or_listexpr).all())
> +
> +VALRANGE_FUNCS = {
> +    'mean':     lambda col, minval, maxval: check(minval <= col.mean() <= maxval),
> +    'mode':     lambda col, minval, maxval: check(minval <= col.mode() <= maxval),
> +    'stddev':   lambda col, minval, maxval: check(minval <= col.std() <= maxval),
> +    'variance': lambda col, minval, maxval: check(minval <= col.var() <= maxval),
> +    'median':   lambda col, minval, maxval: check(minval <= col.median() <= maxval),
> +    'sum':      lambda col, minval, maxval: check(minval <= col.sum() <= maxval),
> +    'count':    lambda col, minval, maxval: check(minval <= col.count() <= maxval),
> +    'abs':      lambda col, minval, maxval: check(
> +        col.between(minval, maxval, inclusive=True).all() )
>
> Why is this one different from the above?
> ------------------------------
>
> In quilt/tools/check_functions.py
> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r140114054>
> :
>
> > +    if 'random_state' not in args:
> +        kwargs['random_state'] = numpy.random.random_sample # pylint:disable=E1101
> +    data = data.sample(*args, **kwargs)
> +
> +def check_column_enum(colrx, lambda_or_listexpr, envs=None):
> +    if envs not in [None, 'default']:
> +        check_column_regexp(colrx, envs[env])
> +    for colname in list(data):
> +        if re.search(colrx, colname):
> +            if callable(lambda_or_listexpr):
> +                check(lambda_or_listexpr(data[colname]))
> +            else:
> +                check(data[colname].isin(lambda_or_listexpr).all())
> +
> +VALRANGE_FUNCS = {
> +    'mean':     lambda col, minval, maxval: check(minval <= col.mean() <= maxval),
>
> I'd move check out of the function, and leave just the lambda ...: minval
> <= ... <= maxval.
>
> (You could be really fancy make it a map of names to methods, e.g. 'mean':
> ColumnType.mean - but I guess that doesn't work for abs? Anyways, just a
> thought - not sure that makes it better.)
> ------------------------------
>
> In quilt/tools/command.py
> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r140114185>
> :
>
> > @@ -14,6 +14,9 @@
>  import sys
>  import time
>  from threading import Thread, Lock
> +import yaml
> +import getpass
>
> Make it alphabetical? I think pylint complains about that (unless you
> disabled that?)
> ------------------------------
>
> In quilt/tools/command.py
> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r140114484>
> :
>
> >                                      compressed_read = 0
>                                      original_last_update = 0
>
>                                  def _progress_cb(count):
> -                                    ctx.compressed_read += count
> -                                    original_read = ctx.compressed_read * original_size // compressed_size
> +                                    Context.compressed_read += count
> +                                    original_read = Content.compressed_read * original_size // compressed_size
>
> Uh... Context, not Content?
>
> This code is kinda impossible to unit test, unfortunately... Though I'd
> expect the linter to catch this.
> ------------------------------
>
> In quilt/tools/command.py
> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r140114661>
> :
>
> > @@ -648,6 +654,19 @@ def install(package, hash=None, version=None, tag=None, force=False):
>              )
>          )
>          pkghash = response.json()['hash']
> +    elif len(hash) < 64:
> +        assert len(hash) >= 6   # short is ok, but only to a point...
>
> This can come from the command-line args. An error is ok - but make sure
> it doesn't assert on user input.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/quiltdata/quilt-compiler/pull/162#pullrequestreview-64140398>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIi4tzixO7cCfDNtExymazwIQfPMnL7ks5skZvKgaJpZM4PN8cU>
> .
>

asah commented on quilt/tools/check_functions.py:
```diff
@@ -0,0 +1,112 @@
+    'sum':      lambda col, minval, maxval: check(minval <= col.sum() <= maxval),
+    'count':    lambda col, minval, maxval: check(minval <= col.count() <= maxval),
+    'abs':      lambda col, minval, maxval: check(
+        col.between(minval, maxval, inclusive=True).all() )
```
DONE - removed, this was experimental and not a requirement for the project.

asah commented:
done - ready for next round.


On Wed, Sep 20, 2017 at 7:09 PM, Adam Sah <asah@midgard.net> wrote:

> No worries from me!  I'm on it.
>
> (via cellphone, random spelling ahead)
>
> On Sep 20, 2017 4:37 PM, "Dima Ryazanov" <notifications@github.com> wrote:
>
>> *@dimaryaz* requested changes on this pull request.
>>
>> Sorry, still found a couple problems.
>>
>> (This is just a really huge change - which is kinda scary...)
>> ------------------------------
>>
>> In quilt/tools/check_functions.py
>> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r140113756>
>> :
>>
>> > +        if re.search(colrx, colname):
>> +            if callable(lambda_or_listexpr):
>> +                check(lambda_or_listexpr(data[colname]))
>> +            else:
>> +                check(data[colname].isin(lambda_or_listexpr).all())
>> +
>> +VALRANGE_FUNCS = {
>> +    'mean':     lambda col, minval, maxval: check(minval <= col.mean() <= maxval),
>> +    'mode':     lambda col, minval, maxval: check(minval <= col.mode() <= maxval),
>> +    'stddev':   lambda col, minval, maxval: check(minval <= col.std() <= maxval),
>> +    'variance': lambda col, minval, maxval: check(minval <= col.var() <= maxval),
>> +    'median':   lambda col, minval, maxval: check(minval <= col.median() <= maxval),
>> +    'sum':      lambda col, minval, maxval: check(minval <= col.sum() <= maxval),
>> +    'count':    lambda col, minval, maxval: check(minval <= col.count() <= maxval),
>> +    'abs':      lambda col, minval, maxval: check(
>> +        col.between(minval, maxval, inclusive=True).all() )
>>
>> Why is this one different from the above?
>> ------------------------------
>>
>> In quilt/tools/check_functions.py
>> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r140114054>
>> :
>>
>> > +    if 'random_state' not in args:
>> +        kwargs['random_state'] = numpy.random.random_sample # pylint:disable=E1101
>> +    data = data.sample(*args, **kwargs)
>> +
>> +def check_column_enum(colrx, lambda_or_listexpr, envs=None):
>> +    if envs not in [None, 'default']:
>> +        check_column_regexp(colrx, envs[env])
>> +    for colname in list(data):
>> +        if re.search(colrx, colname):
>> +            if callable(lambda_or_listexpr):
>> +                check(lambda_or_listexpr(data[colname]))
>> +            else:
>> +                check(data[colname].isin(lambda_or_listexpr).all())
>> +
>> +VALRANGE_FUNCS = {
>> +    'mean':     lambda col, minval, maxval: check(minval <= col.mean() <= maxval),
>>
>> I'd move check out of the function, and leave just the lambda ...:
>> minval <= ... <= maxval.
>>
>> (You could be really fancy make it a map of names to methods, e.g. 'mean':
>> ColumnType.mean - but I guess that doesn't work for abs? Anyways, just a
>> thought - not sure that makes it better.)
>> ------------------------------
>>
>> In quilt/tools/command.py
>> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r140114185>
>> :
>>
>> > @@ -14,6 +14,9 @@
>>  import sys
>>  import time
>>  from threading import Thread, Lock
>> +import yaml
>> +import getpass
>>
>> Make it alphabetical? I think pylint complains about that (unless you
>> disabled that?)
>> ------------------------------
>>
>> In quilt/tools/command.py
>> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r140114484>
>> :
>>
>> >                                      compressed_read = 0
>>                                      original_last_update = 0
>>
>>                                  def _progress_cb(count):
>> -                                    ctx.compressed_read += count
>> -                                    original_read = ctx.compressed_read * original_size // compressed_size
>> +                                    Context.compressed_read += count
>> +                                    original_read = Content.compressed_read * original_size // compressed_size
>>
>> Uh... Context, not Content?
>>
>> This code is kinda impossible to unit test, unfortunately... Though I'd
>> expect the linter to catch this.
>> ------------------------------
>>
>> In quilt/tools/command.py
>> <https://github.com/quiltdata/quilt-compiler/pull/162#discussion_r140114661>
>> :
>>
>> > @@ -648,6 +654,19 @@ def install(package, hash=None, version=None, tag=None, force=False):
>>              )
>>          )
>>          pkghash = response.json()['hash']
>> +    elif len(hash) < 64:
>> +        assert len(hash) >= 6   # short is ok, but only to a point...
>>
>> This can come from the command-line args. An error is ok - but make sure
>> it doesn't assert on user input.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/quiltdata/quilt-compiler/pull/162#pullrequestreview-64140398>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAIi4tzixO7cCfDNtExymazwIQfPMnL7ks5skZvKgaJpZM4PN8cU>
>> .
>>
>

asah commented:
@dimaryaz ready for review

dimaryaz commented on quilt/tools/command.py:
```diff
@@ -667,6 +666,10 @@ def install(package, hash=None, version=None, tag=None, force=False):
                 pkghash = entry['hash']
                 break
+        else:
+            print("could not find hash {hash} for package {owner}/{pkg}.".format(
```
This needs to be `raise CommandException(...)`, to actually indicate an error.

(Also, start messages with uppercase, just for consistency.)

asah replied:
DONE.
